### PR TITLE
Add port for Rasterbar-libtorrent

### PR DESCRIPTION
**Pull request by vktr**
Add port for [Rasterbar-libtorrent](https://github.com/arvidn/libtorrent).

### CODE REVIEW

File: ports/libtorrent/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+vcpkg_extract_source_archive(${ARCHIVE})
+
+vcpkg_apply_patches(
+    SOURCE_PATH ${SOURCE_PATH}
+    PATCHES ${CMAKE_CURRENT_LIST_DIR}/add-datetime-to-boost-libs.patch
+    PATCHES ${CMAKE_CURRENT_LIST_DIR}/add-dbghelp-to-win32-libs.patch
+)
+
+if (NOT VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
+    message(FATAL_ERROR "Only dynamic linking supported.")
```
Comment (atkawa7) [inline ports/libtorrent/portfile.cmake]: @vktr  Thanks for the PR.  Just wondering, I see the portfile on supports dynamic linking. However the following lines show  the possibility for static linking [`https://github.com/arvidn/libtorrent/blob/master/CMakeLists.txt#L255-L281`](https://github.com/arvidn/libtorrent/blob/master/CMakeLists.txt#L255-L281). Could there be a reason for not supporting static linking

File: ports/libtorrent/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+vcpkg_install_cmake()
+
+if (VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
+    # Put shared libraries into the proper directory
+    file(MAKE_DIRECTORY ${CURRENT_PACKAGES_DIR}/bin)
+    file(MAKE_DIRECTORY ${CURRENT_PACKAGES_DIR}/debug/bin)
+
+    file(RENAME ${CURRENT_PACKAGES_DIR}/lib/torrent-rasterbar.dll ${CURRENT_PACKAGES_DIR}/bin/torrent-rasterbar.dll)
+    file(RENAME ${CURRENT_PACKAGES_DIR}/debug/lib/torrent-rasterbar.dll ${CURRENT_PACKAGES_DIR}/debug/bin/torrent-rasterbar.dll)
+else()
```
Comment (atkawa7) [inline ports/libtorrent/portfile.cmake]: @vktr No need for `else()`.

File: ports/libtorrent/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+endif()
+
```
Comment (atkawa7) [inline ports/libtorrent/portfile.cmake]: @vktr  Could you also handle these lines depending on the build i.e static or dynamic. So a user cannot guess. 
[`https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/export.hpp#L64-L78`](https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/export.hpp#L64-L78)

Here is an examples of how you can do that?
[Preprocessor Conditions Examples](https://github.com/Microsoft/vcpkg/search?utf8=%E2%9C%93&q=string%28REPLACE&type=)

File: ports/libtorrent/portfile.cmake (hunk 1 above)
Comment (vktr) [inline ports/libtorrent/portfile.cmake]: I got some errors about not finding the Boost libraries when linking statically. Didn't get very much information more than that.

File: ports/libtorrent/portfile.cmake (hunk 3 above)
Comment (vktr) [inline ports/libtorrent/portfile.cmake]: Could you explain a bit further what you mean? I have used the library for the port I'm making, but I haven't had to define either of those.

File: ports/libtorrent/portfile.cmake (hunk 1 above)
Comment (atkawa7) [inline ports/libtorrent/portfile.cmake]: @vktr Try using the following `-DBoost_USE_STATIC_LIBS=ON` when configuring. See #1338

File: ports/libtorrent/portfile.cmake (hunk 3 above)
Comment (atkawa7) [inline ports/libtorrent/portfile.cmake]: @vktr Suppose a user whats to use the dynamic version of the library he/she has to pass a definition to the compiler `/DTORRENT_BUILDING_SHARED`  if  `cl` or `-DTORRENT_BUILDING_SHARED` with `g++` in order to link correctly. If not defined then the user will get errors `Unresolved symbols ....`. In order for the user not to guess those conditionals we define it for him or her. In this case. After calling cmake install you have to replace 
`#if defined TORRENT_BUILDING_SHARED`  with `#if 1`  if `vcpkg_library_linkage=dynamic`

**Comment by vktr** [comment]
Hi @atkawa7! I've fixed the review comments you had.

**Comment by atkawa7** [comment]
Thanks @vktr . We will have to wait for @ras0219-msft on this one

**Comment by vktr** [comment]
A question - I've been using `find_package` when using vcpkg ports in my own projects. However, this doesn't seem to work for `libtorrent` when I test locally - it doesn't find the library.

It does find OpenSSL and Boost, however, which are also installed with vcpkg. How come?

**Comment by atkawa7** [comment]
@vktr  You could user find_library if find_package does not work

**Comment by vktr** [comment]
@atkawa7 Yes, I noticed that, however it would be nice if it worked. Is there something I can do to make it work before calling the PR finished?

**Comment by atkawa7** [comment]
Thanks @vktr. I am not the maintainer. So we will have to wait on @ras0219-msft and  @alexkaratarakis  on this one.

**Comment by Voskrese** [comment]
**error build**


CMakeFiles\torrent-rasterbar.dir\ed25519\src\sign.cpp.obj 

CMakeFiles\torrent-rasterbar.dir\ed25519\src\verify.cpp.obj E:\tools\vcpkg\installed\x86-windows\lib\boost_system-vc140-mt-1_64.lib E:\tools\vcpkg\installed\x86-windows\lib\boost_date_time-vc140-mt-1_64.lib E:\tools\vcpkg\installed\x86-windows\lib\boost_chrono-vc140-mt-1_64.lib E:\tools\vcpkg\installed\x86-windows\lib\boost_random-vc140-mt-1_64.lib wsock32.lib ws2_32.lib Iphlpapi.lib dbghelp.lib E:\tools\vcpkg\installed\x86-windows\lib\libssl.lib E:\tools\vcpkg\installed\x86-windows\lib\libcrypto.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib 

   тХд╤О╤З╤Д╤А╤Е╨Д╤С┬а ╤Б╤И╤Б╤Л╤И╤О╨Д╤Е╤К╤А torrent-rasterbar.lib ╤И ╤О╤Б┬╖╤Е╤К╨Д torrent-rasterbar.exp

ut_pex.cpp.obj : error LNK2001: ╤Н╤Е╨Б╤А╤З╨Б╤Е┬░╤Е╤Н╤НтИЪ╤Й ╤В╤Н╤Е┬░╤Н╤И╤Й ╤С╤И╤М╤В╤О╤Л "_CRYPTO_cleanup_all_ex_data"

ut_metadata.cpp.obj : error LNK2001: ╤Н╤Е╨Б╤А╤З╨Б╤Е┬░╤Е╤Н╤НтИЪ╤Й ╤В╤Н╤Е┬░╤Н╤И╤Й ╤С╤И╤М╤В╤О╤Л "_CRYPTO_cleanup_all_ex_data"

smart_ban.cpp.obj : error LNK2001: ╤Н╤Е╨Б╤А╤З╨Б╤Е┬░╤Е╤Н╤НтИЪ╤Й ╤В╤Н╤Е┬░╤Н╤И╤Й ╤С╤И╤М╤В╤О╤Л "_CRYPTO_cleanup_all_ex_data"

**... more error**

socket_type.cpp.obj : error LNK2019: ╤С╤СтИЪ╤Л╤К╤А ╤Н╤А ╤Н╤Е╨Б╤А╤З╨Б╤Е┬░╤Е╤Н╤НтИЪ╤Й ╤В╤Н╤Е┬░╤Н╤И╤Й ╤С╤И╤М╤В╤О╤Л _sk_num ╤В ╨З╤Ф╤Н╤К╨О╤И╤И "public: bool __thiscall boost::asio::ssl::rfc2818_verification::operator()(bool,class boost::asio::ssl::verify_context &)const " (??Rrfc2818_verification@ssl@asio@boost@@QBE_N_NAAVverify_context@123@@Z)

torrent.cpp.obj : error LNK2001: ╤Н╤Е╨Б╤А╤З╨Б╤Е┬░╤Е╤Н╤НтИЪ╤Й ╤В╤Н╤Е┬░╤Н╤И╤Й ╤С╤И╤М╤В╤О╤Л "_sk_num"

socket_type.cpp.obj : error LNK2019: ╤С╤СтИЪ╤Л╤К╤А ╤Н╤А ╤Н╤Е╨Б╤А╤З╨Б╤Е┬░╤Е╤Н╤НтИЪ╤Й ╤В╤Н╤Е┬░╤Н╤И╤Й ╤С╤И╤М╤В╤О╤Л _sk_value ╤В ╨З╤Ф╤Н╤К╨О╤И╤И "public: bool __thiscall boost::asio::ssl::rfc2818_verification::operator()(bool,class boost::asio::ssl::verify_context &)const " (??Rrfc2818_verification@ssl@asio@boost@@QBE_N_NAAVverify_context@123@@Z)

torrent.cpp.obj : error LNK2001: ╤Н╤Е╨Б╤А╤З╨Б╤Е┬░╤Е╤Н╤НтИЪ╤Й ╤В╤Н╤Е┬░╤Н╤И╤Й ╤С╤И╤М╤В╤О╤Л "_sk_value"

torrent-rasterbar.dll : fatal error LNK1120: ╤Н╤Е╨Б╤А╤З╨Б╤Е┬░╤Е╤Н╤НтИЪ╤Ч ╤В╤Н╤Е┬░╤Н╤И╤Ч ┬д╤Л╤Е╤М╤Е╤Н╨Д╤О╤В: 17

LINK failed. with 1120
ninja: build stopped: subcommand failed.

**Comment by vktr** [comment]
@Voskrese Strange, it indicates a linking error against OpenSSL but it worked on two of my machines...

**Comment by Voskrese** [comment]
hmmm...
~libssl.lib libcrypto.lib~

dowload https://www.npcglib.org/~stathis/blog/precompiled-openssl/

MSVC 2015 | openssl-1.1.0f-vs2015.7z

libsslMD.lib libcryptoMD.lib
ok

> CMakeFiles\torrent-rasterbar.dir\ed25519\src\verify.cpp.obj E:\tools\vcpkg\installed\x86-windows\lib\boost_system-vc140-mt-1_64.lib E:\tools\vcpkg\installed\x86-windows\lib\boost_date_time-vc140-mt-1_64.lib E:\tools\vcpkg\installed\x86-windows\lib\boost_chrono-vc140-mt-1_64.lib E:\tools\vcpkg\installed\x86-windows\lib\boost_random-vc140-mt-1_64.lib wsock32.lib ws2_32.lib Iphlpapi.lib dbghelp.lib E:\tools\vcpkg\installed\x86-windows\lib\libsslMD.lib E:\tools\vcpkg\installed\x86-windows\lib\libcryptoMD.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib 
> 
>    тХд╤О╤З╤Д╤А╤Е╨Д╤С┬а ╤Б╤И╤Б╤Л╤И╤О╨Д╤Е╤К╤А torrent-rasterbar.lib ╤И ╤О╤Б┬╖╤Е╤К╨Д torrent-rasterbar.exp
> 
> [141/142] cmd.exe /C "cd /D E:\tools\vcpkg\buildtrees\libtorrent\x86-windows-rel && "C:\Program Files\CMake\bin\cmake.exe" -P cmake_install.cmake"
> -- Install configuration: "Release"

**Comment by ras0219-msft** [comment]
Thanks for the PR!

@Voskrese the build works fine on my machine as well. Note that Vcpkg's ports are only tested against openssl 1.0.x, so it wouldn't be surprising if 1.1 causes problems. We do need to start considering what it will take to upgrade to 1.1 though, since 1.0 will end-of-life eventually.

>A question - I've been using find_package when using vcpkg ports in my own projects. However, this doesn't seem to work for libtorrent when I test locally - it doesn't find the library.
>
>It does find OpenSSL and Boost, however, which are also installed with vcpkg. How come?

CMake's `find_package` command[1] is powered by a few different providers, notably:
- Built-in modules shipped with CMake
- Library-provided "config" scripts
- Consumer created FindXYZ.cmake helpers

OpenSSL and Boost both have modules built-in to the base CMake tool. Other libraries such as Eigen ship their own config files (`share/eigen3/Eigen3Config.cmake`). It looks like libtorrent unfortunately has neither.

Since libtorrent uses CMake as its buildsystem, it should be quite easy to generate these config files -- I'd recommend opening an issue or PR on their GitHub requesting/implementing them! Unfortunately, adding them to the project inside Vcpkg is too invasive for us (#616).

That means in the immediate term you'd need to write a FindXYZ.cmake helper yourself if you want to use `find_package()`. However, the implementation of this file is really just a bunch of calls to `find_library()` and `find_path()`, so it's probably easier to just call those directly.

[1] https://cmake.org/cmake/help/v3.8/command/find_package.html

**Comment by Voskrese** [comment]
@ras0219-msft не совсем так, библиотеки стояли с порта vcpkg которые вызвали ошибки, я их на всякий случай даже перекомпилировал, удалив что было. и да, попытка была не одна и я возможно что то упустил с лога, но на память в релиз сборку атачилась debug сборка libcryptomdD.lib и т.д.
а вот как раз путем копирования версии 1.1, все собралось с первого раза

> not exactly, the library with vcpkg port that caused the error, I have them just in case, even perekompiliroval, removing what was. and Yes, the attempt was not one and I may have missed something with log but memory release debug libcryptomdD.lib Assembly atachilas Assembly, etc.
> and that's just by copying the version 1.1, all brought together for the first time

**Pull request merged**